src/common/admin_socket.cc

```diff
@@ -336,23 +336,42 @@ bool AdminSocket::do_accept()
     }
   }
 
-  bool rval = false;
+  bool rval;
+  bufferlist out;
+  rval = execute_command(c, out);
+  if (rval) {
+    uint32_t len = htonl(out.length());
+    int ret = safe_write(connection_fd, &len, sizeof(len));
+    if (ret < 0) {
+      lderr(m_cct) << "AdminSocket: error writing response length "
+          << cpp_strerror(ret) << dendl;
+      rval = false;
+    } else {
+      if (out.write_fd(connection_fd) >= 0)
+        rval = true;
+    }
+  }
 
+  retry_sys_call(::close, connection_fd);
+  return rval;
+}
+
+bool AdminSocket::execute_command(const std::string& cmd, ceph::bufferlist& out)
```

Comment:
nit, could return a `pair<bool, out>` and use structured binding to capture the return values.

also, the commit message could be more specific -- s/Add/Extract/